compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -857,30 +859,52 @@ impl<'a> Parser<'a> {
             components.push(IdentLike(ident_like));
         }
 
-        // FIXME: Make the span more precise.
+        // With proc macros the span can refer to anything. Validate that it refers to something
+        // identical to our float literal, only then we can safely break it into parts.
         let span = self.token.span;
+        let span_is_good = || self.span_to_snippet(span).as_deref() == Ok(float_str).as_deref();
```

Comment:
I worry that this `span_to_snippet` might be too expensive in the parser. Can't we instead look at `self.token.kind`? Of course, this would be called in an uncommon enough codepath that it wouldn't be _too_ bad, but I would prefer to avoid any unnecessary perf expense.

Reply:
The span can point anywhere regardless of the `self.token.kind`, and for an arbitrary span `span.lo + 1` can potentially point e.g. in the middle of a UTF-8 encoded code point.
So, I'm afraid that actually looking into the source map is the only reliable way to determine that you can do arithmetic operations on that span.

>Of course, this would be called in an uncommon enough codepath

Yes, that's what I hope for, `tuple.0.0` is a rare syntactic construction.
Hopefully we'll be able to migrate to more fine-grained tokens ~next year, then the whole `parse_tuple_field_access_expr_float` could be eliminated.